lib/improver/nbhood.py

```diff
@@ -195,6 +198,98 @@ def __repr__(self):
+            y_max = min(cube.shape[0]-1, j+cells_y)
+            summed_array = cube.data
+            total = (summed_array[y_max, x_max] -
+                     summed_array[y_min, x_max]*(y_min >= 0) -
+                     summed_array[y_max, x_min]*(x_min >= 0) +
+                     summed_array[y_min, x_min]*(y_min >= 0 and x_min >= 0))
+            x_min = max(-1, x_min)
+            y_min = max(-1, y_min)
+            area = (y_max - y_min) * (x_max - x_min)
+            return total, area
+
+        yname = cube.coord(axis="y").name()
+        xname = cube.coord(axis="x").name()
+
+        # Calculate displacement factors to find 4-points after flattening
+        # the array.
+        n_rows = len(cube.coord(axis="y").points)
+        n_columns = len(cube.coord(axis="x").points)
+        ymax_xmax_disp = (cells_y*n_columns) + cells_x  # array1
+        ymin_xmax_disp = (-1*(cells_y+1)*n_columns) + cells_x  # array2
+        ymin_xmin_disp = (-1*(cells_y+1)*n_columns) - cells_x - 1  # array3
+        ymax_xmin_disp = (cells_y*n_columns) - cells_x - 1  # array4
+
+        cubelist = iris.cube.CubeList([])
+        for slice_2d in cube.slices([yname, xname]):
+            # Flatten the 2d slice and calculate the sum over the array for
+            # non-edge cases. This is done by creating 4 copies of the
+            # flattened array which are rolled to allign the 4-points which
+            # are needed for the calculation.
+            flattened = slice_2d.data.flatten()
+            array1 = np.roll(flattened, -ymax_xmax_disp)
+            array2 = np.roll(flattened, -ymin_xmax_disp)
+            array3 = np.roll(flattened, -ymin_xmin_disp)
+            array4 = np.roll(flattened, -ymax_xmin_disp)
+            neighbourhood_total = array1 - array2 + array3 - array4
+            neighbourhood_total.resize(n_rows, n_columns)
+
+            # Initialise the neighbourhood size array and calculate
+            # neighbourhood size for non edge cases.
+            neighbourhood_area = np.zeros(neighbourhood_total.shape)
+            neighbourhood_area.fill((2*cells_x+1) * (2*cells_y+1))
+            # Calculate total sum and area of neighbourhood for edge cases.
+            edge_rows = range(cells_x*2) + range(n_rows-2*cells_x, n_rows)
+            edge_columns = range(cells_x*2) + range(n_columns-2*cells_x,
+                                                    n_columns)
+            for j in range(n_rows):
+                for i in (edge_columns):
+                    neighbourhood_total[j, i], neighbourhood_area[j, i] = (
+                        _sum_and_area_for_edge_cases(
+                            slice_2d, cells_x, cells_y, i, j))
+            for i in range(n_columns):
+                for j in (edge_rows):
+                    neighbourhood_total[j, i], neighbourhood_area[j, i] = (
+                        _sum_and_area_for_edge_cases(
+                            slice_2d, cells_x, cells_y, i, j))
```

Comment:
Yes it will be slow for large neighbourhoods. I am not sure how large neighbourhoods are expected to be? 
For a 2d slice of enuk cube running with neighbourhoods of with different number of cells:
Cells: 1  takes:  0.638384103775 seconds
Cells: 5  takes:  2.78943610191 seconds
Cells: 10  takes:  5.45170998573 seconds
Cells: 20  takes:  10.7920219898 seconds
Cells: 50  takes:  26.8557050228 seconds